openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -322,6 +328,43 @@ def test_error_from_mobile_app(self):
         self.request.META = {'HTTP_USER_AGENT': 'open edX Mobile App Version 2.1'}
         self.verify_error(401)
 
+    def test_warn_on_user_change(self):
+        """
+        Verifies that warnings are emitted and custom attributes set if
+        the user changes unexpectedly between request and response.
+        """
+        self.set_up_for_success()
+
+        # But then user changes unexpectedly
+        self.request.user = UserFactory.create()
+
+        with self.assert_logged_with_message(
+            "SafeCookieData user at request '1' does not match user at response: '2' for request path '/'", 'warning'
```

Comment:
nit: this will break if someone needs to add another UserFactory.create(). Probably worth it to interpolate self.user.id and self.request.user.id into the warning string instead of hardcoding 1 and 2